feat: SparkConnect support

SemyonSinchenko opened the pull request:
Work is in progress but the overall design is in the final state:
- subproject `graphframes-connect`
- GraphFrames API in protobuf: graphframes-connect/src/main/protobuf/graphframes.proto
- Proto->GraphFrames->Proto logic: graphframes-connect/src/main/scala/org/apache/spark/sql/graphframes/GraphFramesConnectUtils.scala
- Connect Relation Plugin: graphframes-connect/src/main/scala/org/apache/spark/sql/graphframes/GraphFramesConnect.scala

For JVM part generation is built in to the `build.sbt`; for Python (and possible other clients) `buf` (buf.yaml, buf.gen.yaml)


Close #447

SemyonSinchenko commented:
@rjurney Hello!

At the moment PR provides a (mostly) working version of GrahFrames API for PySpark Connect. There are a lot of open questions from my side:
- I would like to change tests to work for both connect and classic but it can be tricky because current tests are very old; at the moment I just copied the whole tests suite and slightly modified it;
- I do not know what is the best way to integrate Connect API to the current PySpark API: at the moment I implemented everything as a separate `GraphFrameConnect` class that has (almost) the same API like the `GraphFrame`;

How to try it: run the script in `dev` folder. Connect to the Spark Connect Server with `SparkSession.builder.remote("sc://localhost:15002").getOrCreate()` and use `graphframes.connect.graphframes_client.GraphFrameConnect`.

rjurney commented:
> @rjurney Hello!
> 
> At the moment PR provides a (mostly) working version of GrahFrames API for PySpark Connect. There are a lot of open questions from my side:
> 
> * I would like to change tests to work for both connect and classic but it can be tricky because current tests are very old; at the moment I just copied the whole tests suite and slightly modified it;
> * I do not know what is the best way to integrate Connect API to the current PySpark API: at the moment I implemented everything as a separate `GraphFrameConnect` class that has (almost) the same API like the `GraphFrame`;
> 
> How to try it: run the script in `dev` folder. Connect to the Spark Connect Server with `SparkSession.builder.remote("sc://localhost:15002").getOrCreate()` and use `graphframes.connect.graphframes_client.GraphFrameConnect`.

My PR #493 converts the unit tests to `pytest` tests. That should help there. My PR is a monster, I will work on breaking it up into pieces tomorrow and that should give you the ability to pull the test-related code into your PR.

It would be nice if there wasn't a completely different API for Spark Connect. How different are the implementations? I'll take a look at the PR tomorrow and see.

This is very cool work, thanks for it!

SemyonSinchenko commented:
> It would be nice if there wasn't a completely different API for Spark Connect. How different are the implementations? I'll take a look at the PR tomorrow and see.

It is the same, just implemented as a separate class (`python/graphframes/connect/graphframe_client.py`)

rjurney commented:
Is it feasible to add connect support to `GraphFrame`? Especially if it could detect a Connect connection and 'just work' or take an argument?

SemyonSinchenko commented:
> Is it feasible to add connect support to GraphFrame? Especially if it could detect a Connect connection and 'just work' or take an argument?

Of course. That is the idea, but the question is how to do it? In Spark 3.x devs made a decision to make two versions of `DataFrame` (`pyspark.sql.DataFrame` and `pyspark.sql.connect.dataframe.DataFrame`). In the Spark 4.x it is changed and there are three versions at the moment:
- `Classic DataFrame`: https://github.com/apache/spark/blob/master/python/pyspark/sql/classic/dataframe.py
- `Connect DataFrame`: https://github.com/apache/spark/blob/master/python/pyspark/sql/connect/dataframe.py
- `DataFrame` that contains dispatch mechanism: https://github.com/apache/spark/blob/master/python/pyspark/sql/dataframe.py

I did not want to touch the existing code in this PR. I would like to finalize the implementation as a standalone `GraphFrameConnect` class that has exactly the same API as an existing `GraphFrame`. And we can deal with an integration in the next iteration.

What do you think about it?
@rjurney

SemyonSinchenko commented:
To be honest, I like the idea of dispatch like in [PySpark](https://github.com/apache/spark/blob/master/python/pyspark/sql/utils.py#L371), but it will require big changes in the current `GraphFrame`, tests, build, etc. I think it would be better to do in `GraphFrames` 1.0 because I would like to slightly change the public API.

For example, due to Spark Connect limitations it maybe tricky to return both `loss` and `DataFrame` from `svdPlusPlus`. I would like to change the signature and do not return the loss by default. The same about PageRank: I would like to make computation of the edge weights optional with default to False.

SemyonSinchenko commented:
**Questions / Topics**

1. Should we store the generated python code in the repository or not? Apache Spark is storing python code but JVM code is generated during the build. At the moment, my PR follows the way of Apache Spark.
2. Should we have a separate `GraphFramesConnect` or should we use runtime dispatch instead? At the moment my implementation offers a separate class, but Apache Spark uses dispatch.
3. Should a GraphFrames Connect be a part of graphframes distribution or should we provide a separate JAR instead?
4. Should be a Python Connect client be a part of graphframes distribution, or should it be an extra (like `pip install graphframes[connect]`)?

rjurney commented:
@SemyonSinchenko thanks for your work here, this is just awesome. I am going to give it a thorough review this weekend, sooner if I can. Will have questions borne of my lack of Scala knowledge as much as anything :)

SemyonSinchenko commented:
@rjurney What is the plan for the review? Will you review this changes, or should I try to ask someone else to review this?

rjurney commented:
@SemyonSinchenko TODAY

rjurney commented:
@SemyonSinchenko can you describe any changes to the user experience of GraphFrames? Today I am using `databricks-connect` via VSCode for PySpark on Databricks and I can't use `GraphFrame.pageRank` because I am on Connect. Will it work with this PR? What about checkpointing code? That doesn't work either. Just wondering.

SemyonSinchenko commented:
> @SemyonSinchenko can you describe any changes to the user experience of GraphFrames? Today I am using `databricks-connect` via VSCode for PySpark on Databricks and I can't use `GraphFrame.pageRank` because I am on Connect. Will it work with this PR? What about checkpointing code? That doesn't work either. Just wondering.

@rjurney 

For classic users nothing should changes, only some minor / questinable things. For example, I introduced the build of JAR to the process of building the graphframes PySpark. Maybe I should remove it. Should I?

For connect users it is different. They should add graphframes-connect JAR to the Spark Connect Server part and add a spark conf like `spark.connect.extensions.relation.classes=org.apache.spark.sql.graphframes.GraphFramesConnect` to configuration of their Spark Connect Server. That is how the connect plugin system is supposed to be used.

Answers:
1. Will it work on Databricks? Most probably not, but I cannot be sure for 100%. The story is that Databricks Spark is a fork of Apache Spark and it does not match 100% to the Apache Spark. As I can assume based on the delta-io code for PySpark Connect, guys from Databricks backported the signature of plugins from the Spark 4.0 to their branch for DBRs 14.x and 15.x. There is a strong reason for that, the same reason why it was changed from 3.x to 4.x. But my code is targeting Apache Spark 3.4.x and 3.5.x with a workaround, related to the shading rules, to avoid the problem with plugins. So, in my understanding, even with my plugin, GraphFrames won't work on DBR 14.x and DBR 15.x (Spark 3.5.x), but I'm not 100% sure. My question is, should we maintain the DBR compatibility? And if so, how to do it without an access to the source code of DBRs? At the moment there is even no documentation about how to add Spark Connect plugins to the Databricks clusters and you should provide the right configuration before the start of runtime (I'm assuming that the only way is by init scripts that have terrible documentation).
2. I workarounded the checkpoint problem by the usage of the `spark.checkpoint.dir` variable. You can see it, for example, in [these lines](https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L312-L320). For classic the behavior is the same.

rjurney commented:
Dearest Databricks friends, how do we make the new GraphFrames Spark
Connect support Databricks 16.X moving forward? Can we coordinate? People
like me need to run the new GraphFrames PageRank via Spark Connect on
Databricks.

Thanks,
Russell

On Tue, Mar 4, 2025 at 10:40 AM Sem ***@***.***> wrote:

> @SemyonSinchenko <https://github.com/SemyonSinchenko> can you describe
> any changes to the user experience of GraphFrames? Today I am using
> databricks-connect via VSCode for PySpark on Databricks and I can't use
> GraphFrame.pageRank because I am on Connect. Will it work with this PR?
> What about checkpointing code? That doesn't work either. Just wondering.
>
> @rjurney <https://github.com/rjurney>
>
> For classic users nothing should changes, only some minor / questinable
> things. For example, I introduced the build of JAR to the process of
> building the graphframes PySpark. Maybe I should remove it. Should I?
>
> For connect users it is different. They should add graphframes-connect JAR
> to the Spark Connect Server part and add a spark conf like
> spark.connect.extensions.relation.classes=org.apache.spark.sql.graphframes.GraphFramesConnect
> to configuration of their Spark Connect Server. That is how the connect
> plugin system is supposed to be used.
>
> Answers:
>
>    1. Will it work on Databricks? Most probably not, but I cannot be sure
>    for 100%. The story is that Databricks Spark is a fork of Apache Spark and
>    it does not match 100% to the Apache Spark. As I can assume based on the
>    delta-io code for PySpark Connect, guys from Databricks backported the
>    signature of plugins from the Spark 4.0 to their branch for DBRs 14.x and
>    15.x. There is a strong reason for that, the same reason why it was changed
>    from 3.x to 4.x. But my code is targeting Apache Spark 3.4.x and 3.5.x with
>    a workaround, related to the shading rules, to avoid the problem with
>    plugins. So, in my understanding, even with my plugin, GraphFrames won't
>    work on DBR 14.x and DBR 15.x (Spark 3.5.x), but I'm not 100% sure. My
>    question is, should we maintain the DBR compatibility? And if so, how to do
>    it without an access to the source code of DBRs? At the moment there is
>    even no documentation about how to add Spark Connect plugins to the
>    Databricks clusters and you should provide the right configuration before
>    the start of runtime (I'm assuming that the only way is by init scripts
>    that have terrible documentation).
>    2. I workarounded the checkpoint problem by the usage of the
>    spark.checkpoint.dir variable. You can see it, for example, in these
>    lines
>    <https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L312-L320>.
>    For classic the behavior is the same.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKJJL2M7CUV4GOOEEOEGD2SXXTDAVCNFSM6AAAAABWV4AXRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJYGU3DONZVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: SemyonSinchenko]*SemyonSinchenko* left a comment
> (graphframes/graphframes#506)
> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>
>
> @SemyonSinchenko <https://github.com/SemyonSinchenko> can you describe
> any changes to the user experience of GraphFrames? Today I am using
> databricks-connect via VSCode for PySpark on Databricks and I can't use
> GraphFrame.pageRank because I am on Connect. Will it work with this PR?
> What about checkpointing code? That doesn't work either. Just wondering.
>
> @rjurney <https://github.com/rjurney>
>
> For classic users nothing should changes, only some minor / questinable
> things. For example, I introduced the build of JAR to the process of
> building the graphframes PySpark. Maybe I should remove it. Should I?
>
> For connect users it is different. They should add graphframes-connect JAR
> to the Spark Connect Server part and add a spark conf like
> spark.connect.extensions.relation.classes=org.apache.spark.sql.graphframes.GraphFramesConnect
> to configuration of their Spark Connect Server. That is how the connect
> plugin system is supposed to be used.
>
> Answers:
>
>    1. Will it work on Databricks? Most probably not, but I cannot be sure
>    for 100%. The story is that Databricks Spark is a fork of Apache Spark and
>    it does not match 100% to the Apache Spark. As I can assume based on the
>    delta-io code for PySpark Connect, guys from Databricks backported the
>    signature of plugins from the Spark 4.0 to their branch for DBRs 14.x and
>    15.x. There is a strong reason for that, the same reason why it was changed
>    from 3.x to 4.x. But my code is targeting Apache Spark 3.4.x and 3.5.x with
>    a workaround, related to the shading rules, to avoid the problem with
>    plugins. So, in my understanding, even with my plugin, GraphFrames won't
>    work on DBR 14.x and DBR 15.x (Spark 3.5.x), but I'm not 100% sure. My
>    question is, should we maintain the DBR compatibility? And if so, how to do
>    it without an access to the source code of DBRs? At the moment there is
>    even no documentation about how to add Spark Connect plugins to the
>    Databricks clusters and you should provide the right configuration before
>    the start of runtime (I'm assuming that the only way is by init scripts
>    that have terrible documentation).
>    2. I workarounded the checkpoint problem by the usage of the
>    spark.checkpoint.dir variable. You can see it, for example, in these
>    lines
>    <https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L312-L320>.
>    For classic the behavior is the same.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAKJJL2M7CUV4GOOEEOEGD2SXXTDAVCNFSM6AAAAABWV4AXRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJYGU3DONZVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

SemyonSinchenko commented:
@rjurney JFYI. Because I know about the breaking change in connect plugins system from Apache Spark 3.5.x to 4.0.x (the change, that in my understanding is already in the DBR 14.x and 15.x), I separated all the plugins logic from the plugin itself: the plugin itself is less than [five logical lines of code](https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/graphframes-connect/src/main/scala/org/apache/spark/sql/graphframes/GraphFramesConnect.scala) to simplify the migration (and DBR shim) as much as possible. In Apache Spark 4.0.x the signature is changed to `Optional<LogicalPlan> transform(byte[] relation, SparkConnectPlanner planner);` and my workaround with shading of `GraphFrames` proto message is not needed anymore.

rjurney commented:
Let me include Jules and reduce the folks to not overload folks too much.

Saying this, I think we need to figure out how to make GraphFrames work
with Spark Connect first as once we solve that, it should be relatively
straight forward to have it work on Databricks Connect.

Can we start by debugging and identifying the issues with GraphFrames and
Spark Connect?  Jules can you help take point on this and then include
various engineers when we can identify root cause?



Denny Lee
***@***.***

***@***.***>



On Tue, Mar 4, 2025 at 11:35 Russell Jurney ***@***.***>
wrote:

> Dearest Databricks friends, how do we make the new GraphFrames Spark
> Connect support Databricks 16.X moving forward? Can we coordinate? People
> like me need to run the new GraphFrames PageRank via Spark Connect on
> Databricks.
>
> Thanks,
> Russell
>
> On Tue, Mar 4, 2025 at 10:40 AM Sem ***@***.***> wrote:
>
>> @SemyonSinchenko <https://github.com/SemyonSinchenko> can you describe
>> any changes to the user experience of GraphFrames? Today I am using
>> databricks-connect via VSCode for PySpark on Databricks and I can't use
>> GraphFrame.pageRank because I am on Connect. Will it work with this PR?
>> What about checkpointing code? That doesn't work either. Just wondering.
>>
>> @rjurney <https://github.com/rjurney>
>>
>> For classic users nothing should changes, only some minor / questinable
>> things. For example, I introduced the build of JAR to the process of
>> building the graphframes PySpark. Maybe I should remove it. Should I?
>>
>> For connect users it is different. They should add graphframes-connect
>> JAR to the Spark Connect Server part and add a spark conf like
>> spark.connect.extensions.relation.classes=org.apache.spark.sql.graphframes.GraphFramesConnect
>> to configuration of their Spark Connect Server. That is how the connect
>> plugin system is supposed to be used.
>>
>> Answers:
>>
>>    1. Will it work on Databricks? Most probably not, but I cannot be
>>    sure for 100%. The story is that Databricks Spark is a fork of Apache Spark
>>    and it does not match 100% to the Apache Spark. As I can assume based on
>>    the delta-io code for PySpark Connect, guys from Databricks backported the
>>    signature of plugins from the Spark 4.0 to their branch for DBRs 14.x and
>>    15.x. There is a strong reason for that, the same reason why it was changed
>>    from 3.x to 4.x. But my code is targeting Apache Spark 3.4.x and 3.5.x with
>>    a workaround, related to the shading rules, to avoid the problem with
>>    plugins. So, in my understanding, even with my plugin, GraphFrames won't
>>    work on DBR 14.x and DBR 15.x (Spark 3.5.x), but I'm not 100% sure. My
>>    question is, should we maintain the DBR compatibility? And if so, how to do
>>    it without an access to the source code of DBRs? At the moment there is
>>    even no documentation about how to add Spark Connect plugins to the
>>    Databricks clusters and you should provide the right configuration before
>>    the start of runtime (I'm assuming that the only way is by init scripts
>>    that have terrible documentation).
>>    2. I workarounded the checkpoint problem by the usage of the
>>    spark.checkpoint.dir variable. You can see it, for example, in these
>>    lines
>>    <https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L312-L320>.
>>    For classic the behavior is the same.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAKJJL2M7CUV4GOOEEOEGD2SXXTDAVCNFSM6AAAAABWV4AXRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJYGU3DONZVG4>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>> [image: SemyonSinchenko]*SemyonSinchenko* left a comment
>> (graphframes/graphframes#506)
>> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>
>>
>> @SemyonSinchenko <https://github.com/SemyonSinchenko> can you describe
>> any changes to the user experience of GraphFrames? Today I am using
>> databricks-connect via VSCode for PySpark on Databricks and I can't use
>> GraphFrame.pageRank because I am on Connect. Will it work with this PR?
>> What about checkpointing code? That doesn't work either. Just wondering.
>>
>> @rjurney <https://github.com/rjurney>
>>
>> For classic users nothing should changes, only some minor / questinable
>> things. For example, I introduced the build of JAR to the process of
>> building the graphframes PySpark. Maybe I should remove it. Should I?
>>
>> For connect users it is different. They should add graphframes-connect
>> JAR to the Spark Connect Server part and add a spark conf like
>> spark.connect.extensions.relation.classes=org.apache.spark.sql.graphframes.GraphFramesConnect
>> to configuration of their Spark Connect Server. That is how the connect
>> plugin system is supposed to be used.
>>
>> Answers:
>>
>>    1. Will it work on Databricks? Most probably not, but I cannot be
>>    sure for 100%. The story is that Databricks Spark is a fork of Apache Spark
>>    and it does not match 100% to the Apache Spark. As I can assume based on
>>    the delta-io code for PySpark Connect, guys from Databricks backported the
>>    signature of plugins from the Spark 4.0 to their branch for DBRs 14.x and
>>    15.x. There is a strong reason for that, the same reason why it was changed
>>    from 3.x to 4.x. But my code is targeting Apache Spark 3.4.x and 3.5.x with
>>    a workaround, related to the shading rules, to avoid the problem with
>>    plugins. So, in my understanding, even with my plugin, GraphFrames won't
>>    work on DBR 14.x and DBR 15.x (Spark 3.5.x), but I'm not 100% sure. My
>>    question is, should we maintain the DBR compatibility? And if so, how to do
>>    it without an access to the source code of DBRs? At the moment there is
>>    even no documentation about how to add Spark Connect plugins to the
>>    Databricks clusters and you should provide the right configuration before
>>    the start of runtime (I'm assuming that the only way is by init scripts
>>    that have terrible documentation).
>>    2. I workarounded the checkpoint problem by the usage of the
>>    spark.checkpoint.dir variable. You can see it, for example, in these
>>    lines
>>    <https://github.com/SemyonSinchenko/graphframes/blob/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L312-L320>.
>>    For classic the behavior is the same.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/graphframes/graphframes/pull/506#issuecomment-2698567757>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAAKJJL2M7CUV4GOOEEOEGD2SXXTDAVCNFSM6AAAAABWV4AXRGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOJYGU3DONZVG4>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

rjurney commented:
@SemyonSinchenko so is it ready for review then?

SemyonSinchenko commented:
> @SemyonSinchenko so is it ready for review then?

@rjurney Yes, it is. I still need to resolve merge conflicts, but 99% of them are related to the tests, not to the logic. Also one additional method will be introduced to wrap also a recently added power iteration clustering.

SemyonSinchenko commented:
@rjurney I resolved all the conflicts.

An important changes you should check:
- I added the GraphFrames JAR to the python package to avoid JVM-errors;
- I moved tests to python/tests to avoid storing them inside the package;
- I significantly simplify tests by removing all the artifacts of unittests;

At the moment only part of tests are turned on for Connect because other part of tests is based on the `examples` that are called through `py4j`.

SemyonSinchenko commented on build.sbt:
```diff
@@ -20,56 +24,48 @@ ThisBuild / scalaVersion := scalaVer
+    "-Xmx2048m",
+    "-XX:ReservedCodeCacheSize=384m",
+    "-XX:MaxMetaspaceSize=384m",
+    "--add-opens=java.base/sun.nio.ch=ALL-UNNAMED",
```
@rjurney We are testing with 11 and 17... Java 8 is still maintained in Spark 3.5 but it's support is dropped in the upcoming Spark 4.0. Could you please create an issue with your error and I will take a look. Maybe we should add to the ci matrix Java 8 too.

rjurney replied:
Okay, let me try Java 9 now...

rjurney replied:
Okay, so it was Java 8 not 9 that needs to be tested - that is what Spark supports :) Doing it now...

I can't find any issues. Resolving.

SemyonSinchenko commented on graphframes-connect/src/main/protobuf/graphframes.proto:
```diff
@@ -0,0 +1,137 @@
+syntax = 'proto3';
+
+package org.graphframes.connect.proto;
```
Yes, it uses protobuf and gRPC
![image](https://github.com/user-attachments/assets/cf97cda6-75cc-4342-8a26-4dc4f6031dd9)

rjurney replied:
Very cool, thanks

SemyonSinchenko commented on graphframes-connect/src/main/protobuf/graphframes.proto:
```diff
@@ -0,0 +1,137 @@
+
+message TriangleCount {}
+
+message Triplets {}
```
No, the table was made after the first glance. I was able to implement all of APIs.

SemyonSinchenko commented on python/dev/build_jar.py:
```diff
@@ -0,0 +1,54 @@
+from pathlib import Path
+
+
+def build(spark_version: str = "3.5.4"):
```
It is already included with this PR. Take a look on changes in `pyproject.toml`, it contains the reference to this script and resources. It is the reason why I was able to simplify the CI and the local testing too:
https://github.com/SemyonSinchenko/graphframes/blob/refs/heads/447-spark-connect/python/tests/conftest.py#L46-L54

SemyonSinchenko commented on python/dev/run_connect.py:
```diff
@@ -0,0 +1,121 @@
+    else:
+        print(f"SBT build return an error: {build_sbt.returncode}")
+        print("stdout: ", build_sbt.stdout)
+        print("stderr: ", build_sbt.stderr)
```
But why do we need it? This script is for local testing only, it spins the Spark Connect Server with all the necessary configs and with GraphFrames included. I cannot imagine the case when someone wants to run it outside of the GraphFraes development process...

rjurney replied:
I did't mean here - I just meant in general, do we want to move to `logging` based debugs? Seems it might help someone in Python land log things to a log service?

SemyonSinchenko commented on python/dev/stop_connect.py:
```diff
@@ -0,0 +1,35 @@
+
+    os.chdir(spark_home)
+
+    checkpoint_dir = Path("/tmp/GFTestsCheckpointDir")
```
I added a workaround for Spark Connect. Take a look on changes in Connected Components scala code, that are included to the PR:
https://github.com/SemyonSinchenko/graphframes/blob/refs/heads/447-spark-connect/src/main/scala/org/graphframes/lib/ConnectedComponents.scala#L307-L321

rjurney replied:
Awesome, means a lot! Will that work for PageRank as well?

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+    )
+
+
+class GraphFrame:
```
To be honest, I do not like that idea. For me aliases are anti-pythonic things. From Python Zen ([PEP20](https://peps.python.org/pep-0020/)):
```
There should be one-- and preferably only one --obvious way to do it.
```
Why to have two ways to do the same thing? It just makes API confusing... I would agree only if we deprecate `vertices`.

rjurney replied:
Okay, didn't consider that perspective - I just think a lot of people get confused when `g.nodes` isn't there, me included for a long time. Yet it would inconvenience people if they expect `g.vertices` and it goes away.

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+    @property
+    def edges(self) -> DataFrame:
+        return self._edges
+
```
Technically speaking, yes. Please, see my answer to your previous comment.

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+#
```
Yes

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+            raise TypeError("condition should be string or Column")
+        return _from_java_gf(jdf, self._spark)
+
+    def dropIsolatedVertices(self) -> "GraphFrame":
```
I was thinking, that in the case of loops, this method change `edges` internally, but it doesn't, so I was able to implement it in connect.

rjurney commented:
@SemyonSinchenko regarding the JAR, I think it is right to include it and think we should add it to `MANIFEST.in`.

SemyonSinchenko commented on python/graphframes/connect/proto/graphframes_pb2.py:
```diff
@@ -0,0 +1,73 @@
+_sym_db = _symbol_database.Default()
+
+
+DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(
```
Wait, do not look on this. This code is generated automatically by the `protoc` (`buf`)!

SemyonSinchenko commented on python/graphframes/connect/proto/graphframes_pb2.pyi:
```diff
@@ -0,0 +1,342 @@
+        tol: _Optional[float] = ...,
+    ) -> None: ...
+
+class ParallelPersonalizedPageRank(_message.Message):
```
It is generated by `protoc`

SemyonSinchenko commented on python/graphframes/graphframe.py:
```diff
@@ -218,54 +185,40 @@ def find(self, pattern: str) -> DataFrame:
-        return _from_java_gf(jdf, self._spark)
-
-    def filterEdges(self, condition: Union[str, Column]) -> "GraphFrame":
+    def filterEdges(self, condition: str | Column) -> "GraphFrame":
```
I think it is quite new, but it was backported in all supported versions, so all that you need to do is to add a line:
```python
from __future__ import annotations
```
https://peps.python.org/pep-0563/

SemyonSinchenko commented on python/graphframes/graphframe.py:
```diff
@@ -218,54 +185,40 @@ def find(self, pattern: str) -> DataFrame:
-            raise TypeError("condition should be string or Column")
-        return _from_java_gf(jdf, self._spark)
+
+        return GraphFrame._from_impl(self._impl.filterEdges(condition=condition))
```
In the proposed design, all the exception are the responsibility of the implementation. So, answer is yes, it is handled in impl.

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+    def connectedComponents(
+        self,
+        algorithm: str = "graphframes",
+        checkpointInterval: int = 2,
```
Check the answer above. I added a trick to workaround it.

SemyonSinchenko commented on graphframes-connect/src/main/scala/org/apache/spark/sql/graphframes/GraphFramesConnectUtils.scala:
```diff
@@ -0,0 +1,213 @@
+      case MethodCase.LABEL_PROPAGATION => {
+        graphFrame.labelPropagation.maxIter(apiMessage.getLabelPropagation.getMaxIter).run()
+      }
+      case MethodCase.PAGE_RANK => {
```
Take a look on the lines 123-127:
```scala
        if (pageRankProto.hasMaxIter) {
          pageRank.maxIter(pageRankProto.getMaxIter)
        } else {
          pageRank.tol(pageRankProto.getTol)
        }
```

SemyonSinchenko commented on python/graphframes/connect/graphframe_client.py:
```diff
@@ -0,0 +1,865 @@
+                vertex_col_upd=self._update_after_agg_msgs_expr,
+                agg_msg=self._agg_msg,
+                send2dst=self._send_msg_to_dst,
+                send2src=self._send_msg_to_src,
```
Yes!

SemyonSinchenko commented on python/graphframes/connect/graphframe_client.py:
```diff
@@ -0,0 +1,865 @@
+        self._update_after_agg_msgs_expr = updateAfterAggMsgsExpr
+        return self
+
+    def sendMsgToSrc(self, msgExpr: Column | str) -> Self:
```
Yes!

SemyonSinchenko commented on python/graphframes/classic/graphframe.py:
```diff
@@ -0,0 +1,320 @@
+        v = DataFrame(jdf, self._spark)
+        return (v, loss)
+
+    def triangleCount(self) -> DataFrame:
```
Could you create an issue? Or should we deprecate it?

rjurney replied:
We shouldn't deprecate it, it works if the network doesn't have large degrees. We could maybe catch the right exception and add a helpful debug statement.

https://github.com/graphframes/graphframes/issues/540

SemyonSinchenko commented on .gitignore:
```diff
@@ -36,3 +36,9 @@ python/graphframes.egg-info
 python/docs/_site
+
+# JAR that is build during the installation
+python/graphframes/resources/*
```
I think poetry do it internally when you call `poetry build` because this path is specified in the `pyproject.toml` as path to include. Let me check.

SemyonSinchenko replied:
I added the row to `MANIFEST.in`

rjurney replied:
Okay, cool. Otherwise we should whitelist ... actually I thought I created a `MANIFEST.in` when I poetrized it...

Yes: https://github.com/graphframes/graphframes/blob/master/python/MANIFEST.in

Just add the jar to that file :) I think this is the only real issue.

SemyonSinchenko commented on .github/workflows/python-ci.yml:
```diff
@@ -42,16 +40,17 @@ jobs:
       run: |
-        export SPARK_HOME=$(poetry run python -c "import os; from importlib.util import find_spec; spec = find_spec('pyspark'); print(os.path.join(os.path.dirname(spec.origin)))")
-        ./run-tests.sh
+        poetry run python dev/run_connect.py
```
Because I added a JAR to the python resources.

SemyonSinchenko commented on python/dev/build_jar.py:
```diff
@@ -0,0 +1,54 @@
+    if sbt_build.returncode != 0:
+        print("Error during the build of GraphFrames JAR!")
+        print("stdout: ", sbt_build.stdout)
+        print("stdeerr: ", sbt_build.stderr)
```
Fixed

rjurney commented on graphframes-connect/src/main/scala/org/apache/spark/sql/graphframes/GraphFramesConnectUtils.scala:
```diff
@@ -0,0 +1,213 @@
+
+        aggregateMessages.agg(parseColumnOrExpression(aggregateMessagesProto.getAggCol, planner))
+      }
+      case MethodCase.BFS => {
```
Cool!

rjurney commented on python/pyproject.toml:
```diff
@@ -26,13 +35,15 @@ classifiers = [
 python = ">=3.9 <3.13"
 nose = "1.3.7"
-pyspark = "^3.4"
+pyspark = ">=3.3"
```
@SemyonSinchenko this means that Spark 4 will 'just work'?

SemyonSinchenko replied:
I don't think so, tbh. There are breaking changes, especially in the connect. I would like to have a separate issue to track 4.0-compatibility. Let me fix it.

SemyonSinchenko replied:
Fixed.

rjurney reviewed (approved):
lgtm again
